DRIVERS-1807 Keep existing retry behavior unless CSOT is enabled

ShaneHarvey opened the pull request:
Please complete the following before merging:

- [x] Update changelog.
- [x] Make sure there are generated JSON files from the YAML test files.
- [x] Test changes in at least one language driver.
- [x] Test these changes against all server versions and topologies (including standalone, replica set, sharded clusters, and serverless).

benjirewis commented on source/client-side-operations-timeout/client-side-operations-timeout.rst:
```diff
@@ -47,8 +47,10 @@ timeoutMS
-milliseconds. The default value is unset. Both unset and an explicit value of
-0 mean infinite, though some client-side timeouts like
+milliseconds. The default value is unset which means this feature is not
+enabled, ie the existing timeout behavior is unchanged (including
```
```suggestion
enabled, i.e. the existing timeout behavior is unchanged (including
```

benjirewis commented on source/retryable-writes/retryable-writes.rst:
```diff
@@ -381,36 +382,35 @@ writes, drivers MUST add a transaction ID to the command. Drivers MUST only
-error should be raised instead as the caller can infer that an attempt was made
-and the second error is likely more relevant (with respect to the current
-topology state).
+from the driver) or the error is labeled "NoWritesPerformed", the error from
```
Is there any other case besides `NoWritesPerformed` where we return the previous error?

ShaneHarvey replied:
Yes, when a retry selects a server that does not support retryable writes or the retry fails with a server selection error.

benjirewis commented on source/retryable-writes/retryable-writes.rst:
```diff
@@ -443,61 +443,65 @@ Consider the following pseudo-code:
-      throw secondError;
+      /* CSOT is enabled and the operation has timed out. */
+      if (timeoutMS != null && is_expired(timeoutMS) {
+        throw previousError;
```
Do we have a spec test for this block?

ShaneHarvey replied:
Yes we do, it's part of the CSOT. The "timeoutMS applies to whole operation, not individual attempts" tests: https://github.com/mongodb/specifications/blob/b3e06831a09ae72d94da79f428d8941ec7221f7f/source/client-side-operations-timeout/etc/templates/retryability-timeoutMS.yml.template#L42-L47

benjirewis reviewed (commented):
Might want to add @prestonvasquez since some of this has to do with DRIVERS-2327

benjirewis reviewed (approved):
Thanks!

prestonvasquez reviewed (approved):
LGTM!

ShaneHarvey commented:
I've updated this PR to include similar changes to the retryable reads spec. I've also adjusted some of the CSOT spec tests to be less flakey (mainly by increasing timeouts and using w:1 instead of w:majority). Ready for another look.

isabelatkinson reviewed (approved):
lgtm!

benjirewis reviewed (approved):
CSOT test changes LGTM and pass in the Go driver.
